### PR TITLE
Rewrite diet to use knapsack solver.

### LINKED ISSUE #302: use melange when profitable to do so
The thing that makes this spicy is trying to turn it into more hobo consumables. We could try to add mulligan stew/hodgman's blanket to make it easier to fill. 1 pantsgiving fullness + 1 melange + distension pill = slider, which is something I've been abusing in postloop aftercore. We could also start bringing teas into the calculation to make it easier to create sliders, but at that point we're getting crazy.

### LINKED ISSUE #422: Spleen fills insufficiently when optimal spleen item has size > 1

### PR DESCRIPTION
**Pull request by phulin**
Resolves #422
Resolves #302
Probably resolves other things

### CODE REVIEW

File: src/diet.ts [hunk 1]
```diff
@@ -80,33 +97,28 @@ function drinkSafe(qty: number, item: Item) {
   }
   if (!drink(qty, item)) throw "Failed to drink safely";
   if (item.inebriety === 1 && prevDrunk === qty + myInebriety() - 1) {
-    // sometimes mafia does not track the mime army shot glass property
+    // sometimes mafia does not track the mime army shotglass property
```
Comment (horrible-little-slime) [inline src/diet.ts:100]: this is the single most important change

**Review by horrible-little-slime — COMMENTED** [review]
Further commentary will come alongside further testing

File: src/diet.ts [hunk 2]
```diff
@@ -392,16 +315,164 @@ export function runDiet(): void {
...
+    new MenuItem($item`buzzing mushroom wine`),
+    new MenuItem($item`complex mushroom wine`),
+    new MenuItem($item`overpowering mushroom wine`),
+    new MenuItem($item`smooth mushroom wine`),
+    new MenuItem($item`swirling mushroom wine`),
+    new MenuItem($item`Hodgman's blanket`),
+
+    // FOOD
+    new MenuItem($item`jumping horseradish`, {
+      maximum: 1,
```
Comment (horrible-little-slime) [inline src/diet.ts:400]: You said that it filled you up on horseradish? That is probably a bad sign given this

File: src/diet.ts [hunk 3]
```diff
@@ -392,16 +315,164 @@ export function runDiet(): void {
...
+    new MenuItem($item`splendid martini`),
+    new MenuItem($item`punch-drunk punch`),
+    new MenuItem($item`blood-red mushroom wine`),
+    new MenuItem($item`buzzing mushroom wine`),
+    new MenuItem($item`complex mushroom wine`),
+    new MenuItem($item`overpowering mushroom wine`),
+    new MenuItem($item`smooth mushroom wine`),
+    new MenuItem($item`swirling mushroom wine`),
+    new MenuItem($item`Hodgman's blanket`),
+
```
Comment (horrible-little-slime) [inline src/diet.ts:397]: grimlets, spookypockets

File: src/diet.ts [hunk 4]
```diff
@@ -129,12 +144,35 @@ function useIfUnused(item: Item, prop: string | boolean, maxPrice: number) {
...
+    inebrietyLimit() +
+    (have($item`synthetic dog hair pill`) && !get("_syntheticDogHairPillUsed") ? 1 : 0) +
+    (mallPrice($item`spice melange`) < 50 * MPA ? 3 : 0);
+  const spleenAvailable = spleenLimit() + (3 - get("currentMojoFilters"));
+  return (
+    estimatedTurns() +
+    8 * Math.max(0, fullnessAvailable - myFullness()) +
+    8 * Math.max(0, inebrietyAvailable - myInebriety()) +
+    2 * Math.max(0, spleenAvailable - mySpleenUse()) +
+    (have($familiar`Pocket Professor`) && !get("_thesisDelivered") ? 11 : 0)
```
Comment (horrible-little-slime) [inline src/diet.ts:162]: thumb ring multiplier for organ & prof turns--should we not just add thesis turns to the estimatedTurns() function?

File: src/diet.ts [hunk 5]
```diff
@@ -147,140 +185,49 @@ function getBestSpleenItems() {
...
     const synthTurns = haveEffect($effect`Synthesis: Greed`);
     const spleenTotal = spleenLimit() - mySpleenUse();
-    const adventuresPerItem = adventureGain(spleenItem);
+    const adventuresPerSpleen = getAverageAdventures(spleenItem) / spleenItem.spleen;
     // when not barfing, only get synth for estimatedTurns() turns (ignore adv gain)
-    const spleenAdvsGained = globalOptions.noBarf ? 0 : 1.04 * adventuresPerItem * spleenTotal;
+    const spleenAdvsGained = globalOptions.noBarf ? 0 : 1.04 * adventuresPerSpleen * spleenTotal;
     const spleenSynth = Math.ceil(
-      (spleenAdvsGained + estimatedTurns() - synthTurns) / (30 + 1.04 * adventuresPerItem)
+      (spleenAdvsGained + estimatedTurns() - synthTurns) / (30 + 1.04 * adventuresPerSpleen)
```
Comment (horrible-little-slime) [inline src/diet.ts:206]: if we eat a mix of sliders and nonsliders (or drink a combination of pickles and nonpickles), would it be better to use estimatedTurnsWithOrgans?

File: src/diet.ts [hunk 6]
```diff
@@ -147,140 +185,49 @@ function getBestSpleenItems() {
...
-      itemAmount($item`mime army shotglass`) > 0 &&
-      availableAmount($item`astral pilsner`) > 0
-    ) {
+function fillShotglass() {
+  if (!get("_mimeArmyShotglassUsed") && have($item`mime army shotglass`)) {
+    if (have($item`astral pilsner`)) {
       drinkSafe(1, $item`astral pilsner`);
+    } else {
+      equip($item`tuxedo shirt`);
+      drinkSafe(1, $item`splendid martini`);
```
Comment (horrible-little-slime) [inline src/diet.ts:225]: is the martini pretty much always going to be our best bet? Do we envision a scenario where a turkey booze or a sacramento wins out here?

File: src/diet.ts [hunk 7]
```diff
@@ -350,12 +297,10 @@ export function runDiet(): void {
...
 
   if (globalOptions.ascending) useIfUnused($item`borrowed time`, "_borrowedTimeUsed", 5 * MPA);
+}
 
-  fillSomeSpleen();
-  fillStomach();
-  fillLiver();
-
-  if (!get("_distentionPillUsed") && 1 <= myInebriety()) {
+function pillCheck(): void {
```
Comment (horrible-little-slime) [inline src/diet.ts:302]: is it worth adding the "just use a map if you have them" thing to this PR, or should we add that just after?

File: src/diet.ts [hunk 8]
```diff
@@ -392,30 +337,256 @@ export function runDiet(): void {
...
+
+    // HELPERS
+    ...[...stomachLiverCleaners.keys()].map((item) => new MenuItem(item, { maximum: "auto" })),
+    new MenuItem($item`distention pill`, { maximum: "auto" }),
+    new MenuItem($item`cuppa Voraci tea`, { maximum: "auto" }),
+    ...helpers.map((item) => new MenuItem(item)),
+    new MenuItem($item`pocket wish`, { maximum: 1, wishEffect: $effect`Refined Palate` }),
+  ];
+
+  const haveMayo = getWorkshed() === $item`portable Mayo Clinic`;
```
Comment (horrible-little-slime) [inline src/diet.ts:394]: I'll add have() and installed() to libram

File: src/diet.ts [hunk 9]
```diff
@@ -392,30 +337,256 @@ export function runDiet(): void {
...
+        maximum: availableAmount($item`astral pilsner`),
+      })
+    );
   }
 
-  const mojoFilterCount = 3 - get("currentMojoFilters");
-  acquire(mojoFilterCount, $item`mojo filter`, valuePerSpleen(bestSpleenItem), false);
-  if (have($item`mojo filter`)) {
-    use(Math.min(mojoFilterCount, availableAmount($item`mojo filter`)), $item`mojo filter`);
+  if (myLevel() > 26) {
```
Comment (horrible-little-slime) [inline src/diet.ts:422]: >=, right?

File: src/diet.ts (hunk 5 above)
Comment (phulin) [reply src/diet.ts:206]: yep!

File: src/diet.ts (hunk 4 above)
Comment (phulin) [reply src/diet.ts:162]: yep

File: src/diet.ts (hunk 6 above)
Comment (phulin) [reply src/diet.ts:225]: probably, but leaving outside scope for now

File: src/diet.ts (hunk 7 above)
Comment (phulin) [reply src/diet.ts:302]: after

File: src/diet.ts (hunk 9 above)
Comment (phulin) [reply src/diet.ts:422]: yep

File: src/diet.ts (hunk 3 above)
Comment (phulin) [reply src/diet.ts]: done

File: src/diet.ts (hunk 2 above)
Comment (phulin) [reply src/diet.ts]: Reworked this code

File: src/diet.ts [hunk 10]
```diff
@@ -392,16 +315,164 @@ export function runDiet(): void {
...
+
+  const liverStomachCleaners = $items`spice melange, distention pill, synthetic dog hair pill, cuppa Voraci tea, cuppa Sobrie tea`;
+  const helpers = [Mayo.flex, $item`Special Seasoning`, saladFork, frostyMug];
+  const haveMayo = getWorkshed() === $item`portable Mayo Clinic`;
+  const horseradishLength = haveMayo ? 100 : 50;
+  // Adjust value of horseradish if we have mayo, since we're using Mayozapine instead of Mayoflex.
+  const horseradishValue =
+    horseradishLength * baseMeat +
+    Math.min(horseradishLength, embezzlers) * 750 -
+    (haveMayo ? MPA : 0);
```
[the remark this answers is not in the document]
Comment (phulin) [reply src/diet.ts]: Adjustment for mayozapine vs mayoflex.

**Review by horrible-little-slime — APPROVED** [review]

**Pull request merged**